realm/realm-library/src/main/java/io/realm/HandlerController.java

```diff
@@ -54,20 +58,92 @@
     // sorting orders, soring columns, type (findAll, findFirst, findAllSorted etc.)
     final Map<WeakReference<RealmResults<? extends RealmObject>>, RealmQuery<? extends RealmObject>> asyncRealmResults =
             new IdentityHashMap<WeakReference<RealmResults<? extends RealmObject>>, RealmQuery<? extends RealmObject>>();
+    // Keep a WeakReference to the currently empty RealmObjects obtained asynchronously. We need to keep re-running
+    // the query in the background for each commit, until we got a valid Row (pointer)
+    final Map<WeakReference<RealmObject>, RealmQuery<? extends RealmObject>> emptyAsyncRealmObject =
+            new IdentityHashMap<WeakReference<RealmObject>, RealmQuery<? extends RealmObject>>();
+
+    // keep a reference to the list of sync RealmResults, we'll use it
+    // to deliver type based notification once the shared_group advance
+    final IdentitySet<WeakReference<RealmResults<? extends RealmObject>>> syncRealmResults =
+            new IdentitySet<WeakReference<RealmResults<? extends RealmObject>>>();
 
-    final Map<WeakReference<RealmObject>, RealmQuery<? extends RealmObject>> asyncRealmObjects =
+    final Map<WeakReference<RealmObject>, RealmQuery<? extends RealmObject>> realmObjects =
             new IdentityHashMap<WeakReference<RealmObject>, RealmQuery<? extends RealmObject>>();
 
     public HandlerController(BaseRealm realm) {
         this.realm = realm;
     }
 
+    @Override
+    public boolean handleMessage(Message message) {
+        // Due to how a ConcurrentHashMap iterator is created we cannot be sure that other threads are
+        // aware when this threads handler is removed before they send messages to it. We don't wish to synchronize
+        // access to the handlers as they are the prime mean of notifying about updates. Instead we make sure
+        // that if a message does slip though (however unlikely), it will not try to update a SharedGroup that no
+        // longer exists. `sharedGroupManager` will only be null if a Realm is really closed.
+        if (realm.sharedGroupManager != null) {
+            switch (message.what) {
+                case REALM_CHANGED: {
+                    deleteWeakReferences();
```

Comment:
I would move all this logic into it's own method just like the other switch statements